libtransact/src/protocol/receipt.rs

```diff
@@ -61,6 +62,61 @@ impl From<StateChange> for state::StateChange {
     }
 }
 
+impl TryFrom<TransactionResult> for Vec<StateChange> {
+    type Error = String;
+
+    fn try_from(result: TransactionResult) -> Result<Self, Self::Error> {
+        match result {
+            TransactionResult::Valid { state_changes, .. } => Ok(state_changes),
+            TransactionResult::Invalid { .. } => {
+                Err("cannot convert `Invalid` transaction result to state change".into())
+            }
+        }
+    }
+}
+
+impl TryFrom<TransactionResult> for Vec<state::StateChange> {
+    type Error = String;
+
+    fn try_from(result: TransactionResult) -> Result<Self, Self::Error> {
+        let state_changes: Vec<StateChange> = result.try_into()?;
+        Ok(state_changes
+            .into_iter()
+            .map(state::StateChange::from)
+            .collect())
+    }
+}
+
+impl TryFrom<TransactionReceipt> for Vec<StateChange> {
+    type Error = String;
```

Comment:
We should maybe create an error type for this, "InvalidConversionError" which takes the transaction id in its constructor and provides a transaction_id() method. this could then be used by the caller to construct a better error message.

If so, we should do the same for TransactionResult but maybe use an error without the id.

In both cases, because we want to do some processing related to the error, it is probably best to name the errors explicitly so if we change this in the future, we don't introduce errors which could result for another reason without the ability to detect that change occurred.

Reply:
Added new error types